### PR TITLE
reconstruct config and reader

### PR DESCRIPTION
**Pull request by sunxl1988**
gpu: one P40
faster rcnn(batch size 1): 0.298 <-> 0.336 (previous)
yolov3(batch size 8): 0.1883 <-> 0.201 (previous)

add prefetch for data loader
split config yml into 3 parts (reader, architecture, optimize)
add increment config method

adapt mask rcnn & mask rcnn fpn

### CODE REVIEW

File: configs/architechture/cascade_mask_rcnn.yml [hunk 1]
```diff
@@ -102,24 +89,3 @@ Mask:
     resolution: 14
   mask_post_process:
     name: MaskPostProcess
-
-# Train
```
Comment (jerrywgz) [inline configs/architechture/cascade_mask_rcnn.yml:106]: 先只保留mask_rcnn, mask_fpn yolo三个配置文件吧

File: configs/base/cascade_rcnn_r50_1x.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+architecture: CascadeRCNN
```
Comment (jerrywgz) [inline configs/base/cascade_rcnn_r50_1x.yml:1]: 暂时先去掉

File: configs/example/mask_rcnn_r50_1x.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+save_dir: output
+metric: COCO
+num_classes: 81
+
+TrainReader:
+  inputs_def:
+    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class', 'is_crowd', 'gt_poly']
+  dataset:
+    name: COCODataset
+    dataset_dir: /home/ai/dataset/COCO17
```
Comment (jerrywgz) [inline configs/example/mask_rcnn_r50_1x.yml:16]: fix path

File: configs/reader/faster_rcnn.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+TrainReader:
+  inputs_def:
+    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class', 'is_crowd']
+  dataset:
+    name: COCODataset
+    dataset_dir: /home/ai/dataset/COCO17/
```
Comment (jerrywgz) [inline configs/reader/faster_rcnn.yml:6]: fix path

File: configs/reader/faster_rcnn.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    image_dir: train2017
+    anno_path: annotations/instances_train2017.json
+  sample_transforms:
+  - DecodeImage: {to_rgb: true}
+  - RandomFlipImage: {prob: 0.5}
+  - NormalizeImage: {is_channel_first: false, is_scale: true, mean: [0.485,0.456,0.406], std: [0.229, 0.224,0.225]}
```
Comment (jerrywgz) [inline configs/reader/faster_rcnn.yml:12]: Does false or False matter?

File: ppdet/data/__init__.py [hunk 6]
```diff
@@ -1,19 +1,4 @@
-# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline ppdet/data/__init__.py:1]: keep copyright

File: ppdet/data/source/dataset.py [hunk 7]
```diff
@@ -1,140 +1,68 @@
...
-    """
-
+class DetDataset(Dataset):
     def __init__(self,
                  dataset_dir=None,
                  image_dir=None,
                  anno_path=None,
-                 sample_num=-1,
                  with_background=True,
-                 use_default_label=None,
```
Comment (jerrywgz) [inline ppdet/data/source/dataset.py:44]: keep use_default_label for voc dataset

File: ppdet/data/source/dataset.py [hunk 8]
```diff
@@ -1,140 +1,68 @@
...
-    """
-
+class ImageFolder(DetDataset):
     def __init__(self,
                  dataset_dir=None,
                  image_dir=None,
                  anno_path=None,
-                 sample_num=-1,
                  with_background=True,
-                 use_default_label=None,
```
Comment (jerrywgz) [inline ppdet/data/source/dataset.py:120]: keep it as well

File: tools/eval.py [hunk 9]
```diff
@@ -40,7 +39,15 @@ def parse_args():
     return args
 
 
-def run(FLAGS, cfg):
+def run(FLAGS, cfg, place):
+
+    if FLAGS.use_gpu:
```
Comment (jerrywgz) [inline tools/eval.py:44]: devices_num could be removed for the moment

File: tools/train.py [hunk 10]
```diff
@@ -120,81 +122,71 @@ def run(FLAGS, cfg):
...
+                                                outputs.keys())
+                train_stats.update(outputs)
+                logs = train_stats.log()
+                if iter_id % cfg.log_iter == 0:
+                    strs = 'iter: {}, lr: {:.6f}, {}, time: {:.3f}, eta: {}'.format(
+                        iter_id, curr_lr, logs, time_cost, eta)
+                    logger.info(strs)
+
+        # Save Stage 
+        if fluid.dygraph.parallel.Env().local_rank == 0:
```
Comment (jerrywgz) [inline tools/train.py:169]: ParallelEnv().local_rank

File: configs/architechture/cascade_mask_rcnn.yml (hunk 1 above)
Comment (sunxl1988) [reply configs/architechture/cascade_mask_rcnn.yml:106]: 修改

File: configs/base/cascade_rcnn_r50_1x.yml (hunk 2 above)
Comment (sunxl1988) [reply configs/base/cascade_rcnn_r50_1x.yml:1]: 修改

File: configs/reader/faster_rcnn.yml (hunk 4 above)
Comment (sunxl1988) [reply configs/reader/faster_rcnn.yml:6]: 修改

File: ppdet/data/source/dataset.py (hunk 7 above)
Comment (sunxl1988) [reply ppdet/data/source/dataset.py:44]: 父类不用写这个，在VOC的子类里写就行

File: tools/eval.py (hunk 9 above)
Comment (sunxl1988) [reply tools/eval.py:44]: 修改

File: tools/train.py (hunk 10 above)
Comment (sunxl1988) [reply tools/train.py:169]: 修改

File: configs/base/mask_rcnn_r50_1x.yml [hunk 11]
```diff
@@ -9,7 +9,7 @@ log_smooth_window: 20
 save_dir: output
 metric: COCO
 num_classes: 81
-load_static_weights: true
+load_static_weights: false
```
Comment (jerrywgz) [inline configs/base/mask_rcnn_r50_1x.yml:12]: 这里的base也还是需要设置为True，pretrain model是静态图的权重
